File: examples/rust/examples/indexer_authz_v1_query_messages.rs
//! xion.indexer.authz.v1 Query Messages — Protobuf Examples
//!
//! This file demonstrates how to construct, encode, and decode
//! query messages from the Authz Indexer Module (xion.indexer.authz.v1).
//!
//! Queries covered:
//! - QueryGrants: Get grants between specific granter and grantee
//! - QueryGranterGrants: Get all grants issued by a granter
//! - QueryGranteeGrants: Get all grants received by a grantee
//!
//! Note: This is an indexer module - it provides efficient queries for
//! authorization grants that would otherwise require scanning the entire state.
//!
//! **IMPORTANT**: This example requires the `xion_indexer_authz_v1` module to be
//! exported by the xion-types crate. If you get an "unresolved import" error,
//! the crate needs to be updated to include the indexer module exports.

use prost::Message;

use xion_types::types::xion_indexer_authz_v1::{
    QueryGranteeGrantsRequest, QueryGranteeGrantsResponse, QueryGranterGrantsRequest,
    QueryGranterGrantsResponse, QueryGrantsRequest, QueryGrantsResponse,
};

use xion_types::types::cosmos_base_query_v1beta1::{PageRequest, PageResponse};

// =============================================================================
// QueryGrants
// =============================================================================
// Get grants between specific granter and grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants
// Type URL: /xion.indexer.authz.v1.QueryGrantsRequest

fn query_grants_examples() {
    println!("=== QueryGrants Examples ===\n");

    // 1. Construct request
    let request = QueryGrantsRequest {
        granter: "xion1granter...".to_string(),
        grantee: "xion1grantee...".to_string(),
        msg_type_url: "/cosmos.bank.v1beta1.MsgSend".to_string(), // Optional filter
        pagination: Some(PageRequest {
            key: vec![],
            offset: 0,
            limit: 10,
            count_total: false,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryGrantsRequest:");
    println!("   granter: {}", request.granter);
    println!("   grantee: {}", request.grantee);
    println!("   msg_type_url: {}", request.msg_type_url);
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryGrantsRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded:");
    println!("   granter: {}", decoded_request.granter);
    println!("   grantee: {}", decoded_request.grantee);
    println!("   msg_type_url: {}", decoded_request.msg_type_url);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains grants: Vec<cosmos.authz.v1beta1.Grant>
    let response = QueryGrantsResponse {
        grants: vec![], // Would contain Grant objects
        pagination: Some(PageResponse {
            next_key: vec![],
            total: 0,
        }),
    };

    println!("4. Response structure:");
    println!("   grants: Vec<cosmos.authz.v1beta1.Grant>");
    println!("   pagination: PageResponse");
    println!("   grants count: {}", response.grants.len());
}

// =============================================================================
// QueryGranterGrants
// =============================================================================
// Get all grants issued by a granter.
// Endpoint: GET /xion/indexer/authz/v1/grants/granter/{granter}
// Type URL: /xion.indexer.authz.v1.QueryGranterGrantsRequest

fn query_granter_grants_examples() {
    println!("\n=== QueryGranterGrants Examples ===\n");

    // 1. Construct request
    let request = QueryGranterGrantsRequest {
        granter: "xion1granter...".to_string(),
        pagination: Some(PageRequest {
            key: vec![],
            offset: 0,
            limit: 20,
            count_total: true,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryGranterGrantsRequest:");
    println!("   granter: {}", request.granter);
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
        println!("   pagination.count_total: {}", pagination.count_total);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryGranterGrantsRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded granter: {}", decoded_request.granter);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains grants: Vec<cosmos.authz.v1beta1.GrantAuthorization>
    let response = QueryGranterGrantsResponse {
        grants: vec![], // Would contain GrantAuthorization objects
        pagination: Some(PageResponse {
            next_key: vec![],
            total: 5,
        }),
    };

    println!("4. Response structure:");
    println!("   grants: Vec<cosmos.authz.v1beta1.GrantAuthorization>");
    println!("   pagination: PageResponse");
    if let Some(ref pagination) = response.pagination {
        println!("   pagination.total: {}", pagination.total);
    }

    // Use case
    println!("\n--- Use Case ---");
    println!("5. Find all permissions granted BY an account:");
    println!("   - Lists all authorizations this account has given to others");
    println!("   - Useful for auditing delegated permissions");
    println!("   - Returns grantee addresses and authorization details");
}

// =============================================================================
// QueryGranteeGrants
// =============================================================================
// Get all grants received by a grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants/grantee/{grantee}
// Type URL: /xion.indexer.authz.v1.QueryGranteeGrantsRequest

fn query_grantee_grants_examples() {
    println!("\n=== QueryGranteeGrants Examples ===\n");

    // 1. Construct request
    let request = QueryGranteeGrantsRequest {
        grantee: "xion1grantee...".to_string(),
        pagination: Some(PageRequest {
            key: vec![],
            offset: 0,
            limit: 20,
            count_total: true,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryGranteeGrantsRequest:");
    println!("   grantee: {}", request.grantee);
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryGranteeGrantsRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded grantee: {}", decoded_request.grantee);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains grants: Vec<cosmos.authz.v1beta1.GrantAuthorization>
    let response = QueryGranteeGrantsResponse {
        grants: vec![], // Would contain GrantAuthorization objects
        pagination: Some(PageResponse {
            next_key: vec![],
            total: 3,
        }),
    };

    println!("4. Response structure:");
    println!("   grants: Vec<cosmos.authz.v1beta1.GrantAuthorization>");
    println!("   pagination: PageResponse");
    if let Some(ref pagination) = response.pagination {
        println!("   pagination.total: {}", pagination.total);
    }

    // Use case
    println!("\n--- Use Case ---");
    println!("5. Find all permissions received BY an account:");
    println!("   - Lists all authorizations granted TO this account");
    println!("   - Shows what actions this account can perform on behalf of others");
    println!("   - Useful for understanding delegated capabilities");
}

// =============================================================================
// Understanding Authz Indexer
// =============================================================================

fn authz_indexer_explanation() {
    println!("\n=== Understanding Authz Indexer ===\n");

    println!("The Authz Indexer provides efficient queries for authorization grants.\n");

    println!("Key Concepts:\n");

    println!("1. Grant:");
    println!("   - An authorization given from granter to grantee");
    println!("   - Allows grantee to execute specific messages on granter's behalf");
    println!("   - Has an expiration time\n");

    println!("2. GrantAuthorization:");
    println!("   - Extended grant info including granter/grantee addresses");
    println!("   - Used in list responses\n");

    println!("3. Common Authorization Types:");
    let auth_types = [
        ("GenericAuthorization", "Allows any message type"),
        ("SendAuthorization", "Allows bank sends with optional limits"),
        ("StakeAuthorization", "Allows staking operations"),
    ];
    for (name, desc) in &auth_types {
        println!("   - {}: {}", name, desc);
    }

    println!("\n4. Why Use the Indexer?");
    println!("   - Standard authz queries require knowing both granter AND grantee");
    println!("   - Indexer allows querying by granter OR grantee alone");
    println!("   - Much more efficient for listing all grants for an address");
}

// =============================================================================
// Run All Examples
// =============================================================================

fn main() {
    println!("╔════════════════════════════════════════════════════════════╗");
    println!("║  xion.indexer.authz.v1 Query Messages — Protobuf Examples  ║");
    println!("╚════════════════════════════════════════════════════════════╝\n");

    query_grants_examples();
    query_granter_grants_examples();
    query_grantee_grants_examples();
    authz_indexer_explanation();

    println!("\n═══════════════════════════════════════════════════════════════");
    println!("Type URL Reference:");
    println!("  QueryGrantsRequest:        /xion.indexer.authz.v1.QueryGrantsRequest");
    println!("  QueryGranterGrantsRequest: /xion.indexer.authz.v1.QueryGranterGrantsRequest");
    println!("  QueryGranteeGrantsRequest: /xion.indexer.authz.v1.QueryGranteeGrantsRequest");
    println!("\nREST Endpoints:");
    println!("  GET /xion/indexer/authz/v1/grants");
    println!("  GET /xion/indexer/authz/v1/grants/granter/{{granter}}");
    println!("  GET /xion/indexer/authz/v1/grants/grantee/{{grantee}}");
    println!("\nNote: This is a query-only indexer module.");
    println!("═══════════════════════════════════════════════════════════════");
}

File: examples/python/indexer_authz_v1_query_messages.py
#!/usr/bin/env python3
"""
xion.indexer.authz.v1 Query Messages — Protobuf Examples

This file demonstrates how to construct, encode, and decode
query messages from the Authz Indexer Module (xion.indexer.authz.v1).

Queries covered:
- QueryGrants: Get grants between specific granter and grantee
- QueryGranterGrants: List all grants issued by a granter
- QueryGranteeGrants: List all grants received by a grantee

The Authz Indexer provides optimized queries for authorization grants,
allowing efficient lookup of who can act on behalf of whom.

Note: This is an indexer module with query-only messages.
To create/revoke grants, use the standard cosmos.authz.v1beta1 module.
"""

import sys
import os

# Add the python types to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python/types'))

from xion.indexer.authz.v1 import query_pb2 as authz_query
from cosmos.base.query.v1beta1 import pagination_pb2
from google.protobuf.json_format import MessageToJson


# =============================================================================
# QueryGrantsRequest / Response
# =============================================================================
# Get grants between a specific granter and grantee.
# Endpoint: GET /xion/indexer/authz/v1/grants
# Type URL: /xion.indexer.authz.v1.QueryGrantsRequest

def query_grants_examples():
    print('=== QueryGrants Examples ===\n')

    # 1. Construct request
    request = authz_query.QueryGrantsRequest()
    request.granter = 'xion1granter...'  # Who gave the authorization
    request.grantee = 'xion1grantee...'  # Who received the authorization
    request.msg_type_url = '/cosmos.bank.v1beta1.MsgSend'  # Optional: filter by message type
    request.pagination.limit = 10
    request.pagination.offset = 0
    request.pagination.count_total = True

    print('1. Constructed QueryGrantsRequest:')
    print(f'   granter: {request.granter}')
    print(f'   grantee: {request.grantee}')
    print(f'   msg_type_url: {request.msg_type_url or "(all types)"}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGrantsRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # 3. Decode request
    decoded_request = authz_query.QueryGrantsRequest()
    decoded_request.ParseFromString(encoded_request)
    print(f'\n3. Decoded granter: {decoded_request.granter}')
    print(f'   Decoded grantee: {decoded_request.grantee}')

    # === Response ===
    print('\n--- Response ---')

    # Note: Response contains Grant objects from cosmos.authz.v1beta1
    response = authz_query.QueryGrantsResponse()
    # Grants would be populated by the chain
    # response.grants contains cosmos.authz.v1beta1.Grant objects

    print('4. Response structure:')
    print('   grants: [cosmos.authz.v1beta1.Grant]')
    print('   pagination: PageResponse')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGrantsResponse')

    # 4. JSON format
    json_str = MessageToJson(request)
    print(f'\n5. Request JSON format:')
    print(f'   {json_str}')

    return request, response


# =============================================================================
# QueryGranterGrantsRequest / Response
# =============================================================================
# List all grants issued by a specific granter.
# Endpoint: GET /xion/indexer/authz/v1/grants/granter/{granter}
# Type URL: /xion.indexer.authz.v1.QueryGranterGrantsRequest

def query_granter_grants_examples():
    print('\n=== QueryGranterGrants Examples ===\n')

    # 1. Construct request
    request = authz_query.QueryGranterGrantsRequest()
    request.granter = 'xion1granter...'  # List all grants FROM this address
    request.pagination.limit = 20
    request.pagination.count_total = True

    print('1. Constructed QueryGranterGrantsRequest:')
    print(f'   granter: {request.granter}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGranterGrantsRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    # Note: Response contains GrantAuthorization objects
    response = authz_query.QueryGranterGrantsResponse()
    # response.grants contains cosmos.authz.v1beta1.GrantAuthorization objects

    print('3. Response structure:')
    print('   grants: [cosmos.authz.v1beta1.GrantAuthorization]')
    print('     - granter: string')
    print('     - grantee: string')
    print('     - authorization: Any (GenericAuth, SendAuth, etc.)')
    print('     - expiration: Timestamp')
    print('   pagination: PageResponse')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGranterGrantsResponse')

    return request, response


# =============================================================================
# QueryGranteeGrantsRequest / Response
# =============================================================================
# List all grants received by a specific grantee.
# Endpoint: GET /xion/indexer/authz/v1/grants/grantee/{grantee}
# Type URL: /xion.indexer.authz.v1.QueryGranteeGrantsRequest

def query_grantee_grants_examples():
    print('\n=== QueryGranteeGrants Examples ===\n')

    # 1. Construct request
    request = authz_query.QueryGranteeGrantsRequest()
    request.grantee = 'xion1grantee...'  # List all grants TO this address
    request.pagination.limit = 20
    request.pagination.count_total = True

    print('1. Constructed QueryGranteeGrantsRequest:')
    print(f'   grantee: {request.grantee}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGranteeGrantsRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = authz_query.QueryGranteeGrantsResponse()

    print('3. Response structure:')
    print('   grants: [cosmos.authz.v1beta1.GrantAuthorization]')
    print('   pagination: PageResponse')
    print(f'   typeUrl: /xion.indexer.authz.v1.QueryGranteeGrantsResponse')

    return request, response


# =============================================================================
# Understanding Authz Grants
# =============================================================================

def authz_explanation():
    print('\n=== Understanding Authz Grants ===\n')

    print('Authz (Authorization) allows one account to act on behalf of another.\n')

    print('Key Concepts:\n')
    print('  1. Granter: The account giving permission')
    print('  2. Grantee: The account receiving permission')
    print('  3. Authorization: What actions are permitted\n')

    print('Common Authorization Types:\n')
    auth_types = [
        ('GenericAuthorization', 'Allows any message of a specific type'),
        ('SendAuthorization', 'Allows sending up to a spend limit'),
        ('StakeAuthorization', 'Allows staking operations (delegate, undelegate, redelegate)'),
    ]

    for auth_type, desc in auth_types:
        print(f'  • {auth_type}:')
        print(f'    {desc}\n')

    print('Use Cases:\n')
    print('  • Fee grants: Let another account pay your fees')
    print('  • Automated staking: Let a bot manage your delegations')
    print('  • Multi-sig alternatives: Grant specific permissions to team members')
    print('  • Abstract accounts: Meta-transactions and gasless UX\n')

    print('Indexer vs Standard Authz:\n')
    print('  • This module provides OPTIMIZED QUERIES for grants')
    print('  • To CREATE grants, use: cosmos.authz.v1beta1.MsgGrant')
    print('  • To REVOKE grants, use: cosmos.authz.v1beta1.MsgRevoke')
    print('  • To EXECUTE with grant, use: cosmos.authz.v1beta1.MsgExec')


# =============================================================================
# Run All Examples
# =============================================================================

def main():
    print('╔════════════════════════════════════════════════════════════╗')
    print('║  xion.indexer.authz.v1 Query Messages — Protobuf Examples  ║')
    print('╚════════════════════════════════════════════════════════════╝\n')

    query_grants_examples()
    query_granter_grants_examples()
    query_grantee_grants_examples()
    authz_explanation()

    print('\n═══════════════════════════════════════════════════════════════')
    print('Type URL Reference (Requests):')
    print('  QueryGrantsRequest:         /xion.indexer.authz.v1.QueryGrantsRequest')
    print('  QueryGranterGrantsRequest:  /xion.indexer.authz.v1.QueryGranterGrantsRequest')
    print('  QueryGranteeGrantsRequest:  /xion.indexer.authz.v1.QueryGranteeGrantsRequest')
    print('\nType URL Reference (Responses):')
    print('  QueryGrantsResponse:         /xion.indexer.authz.v1.QueryGrantsResponse')
    print('  QueryGranterGrantsResponse:  /xion.indexer.authz.v1.QueryGranterGrantsResponse')
    print('  QueryGranteeGrantsResponse:  /xion.indexer.authz.v1.QueryGranteeGrantsResponse')
    print('\nREST Endpoints:')
    print('  GET /xion/indexer/authz/v1/grants')
    print('  GET /xion/indexer/authz/v1/grants/granter/{granter}')
    print('  GET /xion/indexer/authz/v1/grants/grantee/{grantee}')
    print('\nNote: This is an indexer module (query-only).')
    print('Use cosmos.authz.v1beta1 for grant/revoke transactions.')
    print('═══════════════════════════════════════════════════════════════')


if __name__ == '__main__':
    main()
